build: specs sources and tooling to package Clear Containers

gorozco1 opened the pull request:
Package specs to create create clear-containers packages
in OBS for distros like: Fedora, Ubuntu, CentOS and RHEL
Signed-off-by: Geronimo Orozco <geronimo.orozco@intel.com>

chavafg commented:
qa-failed

jodh-intel commented on data/obs-packaging/README.md:
```diff
@@ -0,0 +1,4 @@
+#Clear Containers OBS packages
+
+
+WIP
```
This top-level README really needs some text to explain what this is all about.

I'd add a brief summary explaining what OBS is (including a link to the main site) and why we use it.

gorozco1 replied:
yes.. working on that...

jodh-intel commented:
The file `data/obs-packaging/linux-container/linux-4.9.4.tar.xz` is empty. I'm wondering how best we deal with this file which is pretty big. Rather than storing the kernel sources in-tree, it might be better to allow users to specify the path to a local copy of that file (`configure --with-cc-obs-kernel-source=...` or `make CC_OBS_KERNEL_SOURCE=...` ?), but also have an option to download it maybe.

grahamwhaley commented:
nitpick: subject prefix please :-) https://github.com/01org/cc-oci-runtime/blob/master/CONTRIBUTING.md#patch-format

sboeuf commented:
@gorozco1 any updates on this one ?

gorozco1 commented:
@sboeuf Need to update this with the latest code

sboeuf commented:
@gorozco1 no problem, please remove the do-not-merge label when it is done.

gorozco1 commented:
@jodh-intel yes `data/obs-packaging/linux-container/linux-4.9.4.tar.xz` is empty because is the linux kernel source code. :P

jodh-intel commented:
@gorozco1 - I know what the file is supposed to contain, but I don't think we want to store a zero-length file.

I suggest you:

- Remove the `.tar.xz` files entirely.
- Add a script that calls `spectool -g foo.spec` to actually download the (huge) archives.
- Add a `README.md` explaining what the script is for and how to run it.

jodh-intel commented:
This is looking good. I worry slightly though about the uniqueness of the tags in the template files - could you maybe change `VERSION`, `HASH_TAG`, *etc* to be `@VERSION@`, `@HASH_TAG@`, ... to make it less likely that `sed` will change some other part of these files inadvertently.

dlespiau commented:
We don't usually sign our commits btw.

grahamwhaley commented:
@dlespiau Just to be clear (as I had to think/look...) you mean GPG key sign the commits, not 'Sign-Off-By', yes :-)
Indeed, we've not had any key signed commits before. fmi, any particular reason not to?

dlespiau commented:
We don't need to :) the blockchain-like git history already enforces quite nice cryptographic properties. Usually people sign their git tags though (tags aren't part of the commit chain).

chavafg commented:
qa-passed

jodh-intel commented on data/obs-packaging/attr-dev/attr-dev.spec:
```diff
@@ -0,0 +1,234 @@
+%package -n libpng-dev
+#%if 0%{?centos_version}
+#Prefer: -libpng12-devel
+#%endif
```
Commented out code?

jodh-intel commented on data/obs-packaging/README.rst:
```diff
@@ -0,0 +1,39 @@
+====================================
+
+
+The `Clear Containers`_ packages were made available via `Open Build Service`_ to allow interested parties the opportunity to try out the technology, even if they are not using Clear Linux directly.
```
s/were made/are made/. Also, maybe make "Clear Linux" a link to clearlinux.org?

jodh-intel commented on data/obs-packaging/README.rst:
```diff
@@ -0,0 +1,39 @@
+
+
+The `Clear Containers`_ packages were made available via `Open Build Service`_ to allow interested parties the opportunity to try out the technology, even if they are not using Clear Linux directly.
+Once `cc-oci-runtime`_ or any of its components are updated we release a new package or set of packages.
```
s/updated we/updated, we/

jodh-intel commented on data/obs-packaging/README.rst:
```diff
@@ -0,0 +1,39 @@
+- `clear-containers-image`_
+- `clear-containers-selinux`_
+- `linux-container`_
+- `qemu-lite`_
```
I'd be tempted to add a brief description to each of these components. Something like:

- `clear-containers-image`_: Contains the mini-OS required to run Clear Containers.
- `linux-container`_: The optimised guest kernel required to run Clear Containers.
- `qemu-lite`_: An optimised version of the QEMU hypervisor.

jodh-intel commented on data/obs-packaging/cc-oci-runtime/cc-oci-runtime.spec-template:
```diff
@@ -0,0 +1,162 @@
+/usr/lib/systemd/system/container.target
+/usr/lib/systemd/system/opt-rootfs-proc.mount
+/usr/lib/systemd/system/opt-rootfs-sys.mount
+/usr/lib/systemd/system/opt-rootfs.mount
```
These 3 `opt-*` files got zapped recently in #840.

jodh-intel commented on data/obs-packaging/cc-oci-runtime/cc-oci-runtime.spec-template:
```diff
@@ -0,0 +1,162 @@
+%exclude /usr/lib/systemd/system/container.target
+%exclude /usr/lib/systemd/system/opt-rootfs-proc.mount
+%exclude /usr/lib/systemd/system/opt-rootfs-sys.mount
+%exclude /usr/lib/systemd/system/opt-rootfs.mount
```
I think the `opt-*.mount` files can go now thanks to #840.

jodh-intel commented on data/obs-packaging/cc-oci-runtime/update_runtime.sh:
```diff
@@ -0,0 +1,28 @@
+#!/bin/bash
+# $1 new version defaults to 2.1.4
```
Is that statement still true?

jodh-intel commented on data/obs-packaging/clear-containers-image/README.md:
```diff
@@ -0,0 +1,9 @@
+
+``./update_image.sh``
+
+By default it will generate the specs using the version specified in ../../../versions.txt
```
Maybe add backticks for `version.txt`. It might be simpler to say that the file `versions.txt` at in the root of the repository rather than all the `../../` bits.

jodh-intel commented:
See #852.

jodh-intel commented:
Hi @gorozco1 - I've done a relatively full review now. One final comment: we have a couple of relatively mysterious spec files (`elfutils-devel.spec` and `attr-dev.spec`). As I recall, they were originally added as "fake" packages to handle pulling in the extra build dependencies (I think `attr-dev.spec` was specifically for `qemu-lite`).

I think we need to re-assess the situation here. Do we still need them? Can we find another way? If we are stuck with them, I wonder if we can give them more meaningful name (`clear-containers-fake-package-for-build-depencies` or something ;-)

If you look at the OBS build output, we have a nice set of packages for Ubuntu (for example http://download.opensuse.org/repositories/home:/clearlinux:/preview:/clear-containers-2.1/xUbuntu_16.10/amd64/), but a fairly odd (and large) set of packages for Centos (http://download.opensuse.org/repositories/home:/clearlinux:/preview:/clear-containers-2.1/CentOS_7/x86_64/) and Fedora (http://download.opensuse.org/repositories/home:/clearlinux:/preview:/clear-containers-2.1/Fedora_25/x86_64/).

An obvious question a user/dev is going to ask is "why do we need all these extra packages for RPM-based distros?" I think as a minimum we need to document exactly what these extra spec files are for because we really shouldn't need to be building packages that are already available in their respective distros. Note also that one of the packages is openssl-dev.

gorozco1 commented on data/obs-packaging/cc-oci-runtime/debian.rules:
```diff
@@ -0,0 +1,28 @@
+	dh $@
+
+override_dh_auto_configure:
+	sh ./autogen.sh --disable-static \
```
that's a good idea @jodh-intel for the time being I will update the options and raise an issue related to  qemu-lite build options.

gorozco1 replied:
issue #853 has been created related to this.

chavafg commented:
qa-passed

gorozco1 commented:
@jodh-intel Thanks for reviewing this PR... 
I am not sure if the extra packages are still needed to build cc-oci-runtime and its dependencies. Let me run some checks..

chavafg commented:
qa-passed

amshinde commented on data/obs-packaging/qemu-lite/qemu-lite.spec:
```diff
@@ -9,25 +9,20 @@ Group    : Development/Tools
+BuildRequires : zlib
 BuildRequires : pkgconfig(pixman-1)
 Patch1: configure.patch
 
```
@gorozco1 Had a look at this commit, this looks good.

gorozco1 replied:
@amshinde thanks for your review

gorozco1 commented:
@jodh-intel I addressed all your comments and also removed the "packages not needed" like attr-dev. I'll appreciate if you can review changes in the last two commits.

jodh-intel commented:
Hi @gorozco1 - this is a big change and a very useful one! I can't wait to see this landed as it will simplify and refine our process a lot. Thanks!

lgtm.

jodh-intel commented:
@rcaballeromx - could you take a look at the doc changes on this PR please?

chavafg commented:
qa-passed

chavafg commented:
qa-passed

sboeuf commented:
lgtm

chavafg commented:
qa-passed

jodh-intel commented:
Hi @rcaballeromx - just a reminder that we need a doc review of this important PR.

chavafg commented:
qa-passed

jodh-intel commented:
Hi @gorozco1 - pullapprove is showing that 2 of your commits don't have SOB's.

gorozco1 commented:
@jodh-intel Oh, thanks for heads-up fixed with a rebase

chavafg commented:
qa-passed

jodh-intel commented:
lgtm

rcaballeromx commented on data/obs-packaging/cc-oci-runtime/README.md:
```diff
@@ -0,0 +1,10 @@
+#cc-oci-runtime
+
+This source rpm is the clear-containers runtime used to create clear-containers. 
+To update and generate an rpm spec and dsc to create rpm and deb packages, 
```
I am editing this file but this sentence is undecipherable. Please avoid using so many abbreviations in a sentence. If the script performs multiple operations with outputs, consider using a list. I believe this to be the case.

rcaballeromx commented on data/obs-packaging/clear-containers-image/README.md:
```diff
@@ -0,0 +1,10 @@
+#clear-containers image
+
+This source rpm is the mini-os image used to boot clear-containers. 
+To update and generate an rpm spec and dsc to create rpm and deb packages, 
```
Just as above. This sentence makes no sense. I suggest one of two paths:
1. Explicitly list the actions and results of the script.
2. Change the sentence to a simple high-level description with one action with one result.

gorozco1 commented:
@rcaballeromx updated could you please check

chavafg commented:
qa-failed

chavafg commented:
qa-passed

chavafg commented:
qa-passed

chavafg commented:
qa-passed

rcaballeromx commented on data/obs-packaging/README.rst:
```diff
@@ -0,0 +1,45 @@
+===================================
+
+
+The `Clear Containers`_ packages are made available via `Open Build Service`_ to allow interested parties the opportunity to try out the technology, even if they are not using `Clear Linux` directly.
```
Please fix the line's length.

rcaballeromx commented on data/obs-packaging/README.rst:
```diff
@@ -0,0 +1,45 @@
+Components
+----------
+
+- `cc-oci-runtime`_: The Clear Containers runtime.
```
The components list is still the old one.

rcaballeromx commented on data/obs-packaging/cc-oci-runtime/README.md:
```diff
@@ -0,0 +1,18 @@
+#cc-oci-runtime
+
+This directory contains the sources to update rpm specfiles and debian source
```
... to update the rpm specification files ...

rcaballeromx commented on data/obs-packaging/cc-oci-runtime/README.md:
```diff
@@ -0,0 +1,18 @@
+This directory contains the sources to update rpm specfiles and debian source
+control files to generate ``cc-oci-runtime`` packages for Fedora and Ubuntu.
+
+There is a script to generate specs and dsc automatically, to be pushed to
```
The following script generates specification files and dsc automatically.

rcaballeromx commented on data/obs-packaging/cc-oci-runtime/README.md:
```diff
@@ -0,0 +1,18 @@
+control files to generate ``cc-oci-runtime`` packages for Fedora and Ubuntu.
+
+There is a script to generate specs and dsc automatically, to be pushed to
+Open Build Service (http://openbuildservice.org/) and generate packages for you.
```
Those files can be pushed to [Open Build Service] ( http://openbuildservice.org/) to generate the packages.

rcaballeromx commented on data/obs-packaging/cc-oci-runtime/README.md:
```diff
@@ -0,0 +1,18 @@
+
+``./update_runtime.sh``
+
+This script will generate/update the sources to create ``cc-oci-runtime`` packages.
```
This script generates and updates the sources to create the ``cc-oci-runtime`` packages.

rcaballeromx commented on data/obs-packaging/cc-oci-runtime/README.md:
```diff
@@ -0,0 +1,18 @@
+  * cc-oci-runtime.spec
+
+
+By default it will generate the specs using the version specified in the file
```
This text is still the old version. Please replace with the provided edits.

rcaballeromx commented on data/obs-packaging/clear-containers-image/README.md:
```diff
@@ -0,0 +1,18 @@
+#clear-containers-image
```
See the edits to the previous file and apply them here too.

rcaballeromx commented on data/obs-packaging/clear-containers-image/README.md:
```diff
@@ -0,0 +1,18 @@
+
+``./update_image.sh``
+
+This script will generate/update:
```
Avoid using slashes, they are ambiguous. Do you mean:
This script generates and updates:
OR
This script generates or updates:

gorozco1 replied:
@rcaballeromx  fixed in last commit

rcaballeromx commented on data/obs-packaging/clear-containers-selinux/README.md:
```diff
@@ -0,0 +1,3 @@
+#clear-containers-selinux
```
This file does not reflect the edits provided. Please apply the edits.

rcaballeromx commented on data/obs-packaging/linux-container/README.md:
```diff
@@ -0,0 +1,17 @@
+#linux-container
```
See the previous edits and comments and apply them here as well.

rcaballeromx commented on data/obs-packaging/linux-container/README.md:
```diff
@@ -0,0 +1,17 @@
+
+``./update_kernel.sh``
+
+This scirpt will generate/update teh sources to create ``linux-container`` packages
```
This script generates and updates the sources to create the following ``linux-container`` packages:

gorozco1 commented:
@rcaballeromx could you provide the files via PR to my branch (obs):

https://github.com/gorozco1/cc-oci-runtime

rcaballeromx commented:
You can find the commit here:
https://github.com/gorozco1/cc-oci-runtime/commit/a6c6669af92ffe38be2f6a85cb293c0503c2fe4b

gorozco1 commented:
@rcaballeromx I apply your changes..

chavafg commented:
qa-passed

chavafg commented:
qa-passed

chavafg commented:
qa-failed

gorozco1 commented:
@rcaballeromx I apply the changes you requested on the last review, thanks for the update.

chavafg commented:
qa-passed

chavafg commented:
qa-passed

gorozco1 commented:
@rcaballeromx any updates on this? all changes requested have been applied

chavafg commented:
qa-failed

jodh-intel commented:
Hi @chavafg - please could you re-run the QA checks on this PR?

chavafg commented:
qa-passed
